### PR TITLE
[IndexTable] Allow header text/content alignment and adjust right-aligned, sortable animation

### PR DESCRIPTION
**Pull request by rcaplanshopify**
### WHY are these changes introduced?

Fixes [#407753](https://github.com/Shopify/shopify/issues/407753) 

Once this update is generally available, the headings associated with the table columns modified in [this PR](https://github.com/Shopify/web/pull/84377#discussion_r1115758460) will be updated to use the alignment property, and the linked Web PR will ship.

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: polaris-react/src/components/IndexTable/IndexTable.scss [hunk 1]
```diff
@@ -268,6 +284,32 @@ $loading-panel-height: 53px;
...
+    left: 0;
+    margin-right: 0;
+  }
+}
+
+.TableHeadingSortButton-heading-align-end-previously-sorted {
+  animation: hide-sort-button var(--p-duration-200) var(--p-ease-in);
+
+  .TableHeadingSortIcon {
+    animation: hide-sort-button-icon var(--p-duration-200) var(--p-ease-in);
```
Comment (rcaplanshopify) [inline polaris-react/src/components/IndexTable/IndexTable.scss]: These styles animate the width of the sort icon after an active, right-aligned, sortable column becomes inactive due to a click on another sortable column.

File: polaris-react/src/components/IndexTable/IndexTable.scss [hunk 2]
```diff
@@ -282,6 +324,17 @@ $loading-panel-height: 53px;
...
 
+.TableHeadingSortIcon-heading-align-end {
+  transition: width var(--p-duration-200) var(--p-ease-in) var(--p-duration-100);
+  overflow: hidden;
+  width: 0;
+
+  svg {
+    min-height: var(--p-space-5);
+    min-width: var(--p-space-5);
+  }
```
Comment (rcaplanshopify) [inline polaris-react/src/components/IndexTable/IndexTable.scss]: This prevents scaling of the svg as its container width increases. Without this, it looks like the svg comes in from the top and expands to take up available area as the container width increases.

File: polaris-react/src/components/IndexTable/IndexTable.scss (hunk 1 above)
Comment (alex-page) [reply polaris-react/src/components/IndexTable/IndexTable.scss]: I understand why the animation was added. The previous design having the sort icons up against the text in the title means there is no room on the right. I think the animation adds very little value to the user experience of sorting lists in a power tool.

I am happy to ship this but I think a simpler design or a smaller column width (where right aligned titles are not needed)  would remove the necessity to animate title cells when trying to perform important actions.

File: polaris-react/src/components/IndexTable/IndexTable.scss [hunk 3]
```diff
@@ -268,6 +284,32 @@ $loading-panel-height: 53px;
...
+    .TableHeadingSortIcon-heading-align-end {
+      width: var(--p-space-4);
+    }
+  }
+}
+
+.TableHeadingSortButton-heading-align-end {
+  position: relative;
+  left: var(--p-space-4);
+  margin-right: var(--p-space-4);
```
Comment (alex-page) [inline polaris-react/src/components/IndexTable/IndexTable.scss]: Can we use `translateX` instead? It is much more performant for animations then animating `margin` and `left`. Animating `left` and `margin-right` causes a repaint. Read more here https://web.dev/animations-guide/#triggers

File: polaris-react/src/components/IndexTable/IndexTable.scss [hunk 4]
```diff
@@ -282,6 +324,17 @@ $loading-panel-height: 53px;
     var(--p-duration-100);
 }
 
+.TableHeadingSortIcon-heading-align-end {
+  transition: width var(--p-duration-200) var(--p-ease-in) var(--p-duration-100);
+  overflow: hidden;
+  width: 0;
```
Comment (alex-page) [inline polaris-react/src/components/IndexTable/IndexTable.scss]: Same as above can we animate more performant values then `width`?

**Review by alex-page — APPROVED** [review]
We should fix the CSS animation performance before shipping.

**Review by mrcthms — APPROVED** [review]
Only concern here is the same as Alex's about the animations. I don't think we'll need to animate the width of the icon, we should be able to animate only the opacity of the icon and the `translateX` of the label. Whipped up a super simple Codepen of how it could work: https://codepen.io/mrcthms/pen/oNPZayv

**Comment by rcaplanshopify** [comment]
Thanks for reviewing @mrcthms and @alex-page! I'm going to work on the animation performance optimization today and will let you know when the optimization changes are ready for review.

File: polaris-react/src/components/IndexTable/IndexTable.scss (hunk 4 above)
Comment (rcaplanshopify) [reply polaris-react/src/components/IndexTable/IndexTable.scss]: I was able to achieve the same overall animated effect without the icon manipulation.

**Review by alex-page — APPROVED** [review]

**Pull request merged**